src/org/labkey/test/components/ui/entities/EntityInsertPanel.java

```diff
@@ -196,6 +264,64 @@ protected boolean isVisible(Locator locator)
         }
     }
 
+    public EntityInsertPanel showGrid()
+    {
+        if (!isGridVisible())
+            elementCache().modeSelectListItem("from Grid")
+                    .waitForElement(this, 2000).click();
+        clearElementCache();
+        WebDriverWrapper.waitFor(()-> isGridVisible(),
+                "the grid did bot become visible", 2000);
+        return this;
+    }
+
+    public ResponsiveGrid uploadFileExpectingPreview(File file)
+    {
+        showFileUpload();
+        fileUploadPanel().uploadFile(file);
+        return new ResponsiveGrid.ResponsiveGridFinder(getDriver()).waitFor(this);
+    }
+
+    public EntityInsertPanel showFileUpload()
+    {
+        if (!isFileUploadVisible())
+        {
+            var toggle = elementCache().modeSelectListItem("from File")
+                    .waitForElement(this, 2000);
+            getWrapper().shortWait().until(ExpectedConditions.elementToBeClickable(toggle));
+            toggle.click();
+        }
+        clearElementCache();
+        WebDriverWrapper.waitFor(()-> isFileUploadVisible(),
+                "the file upload panel did bot become visible", 2000);
```

Comment:
No need to clear the cache and wait for the panel if the panel is already in file upload mode. Put all of this under the `if`:
```suggestion
        if (!isFileUploadVisible())
        {
            var toggle = elementCache().modeSelectListItem("from File")
                    .waitForElement(this, 2000);
            getWrapper().shortWait().until(ExpectedConditions.elementToBeClickable(toggle));
            toggle.click();
            clearElementCache();
            WebDriverWrapper.waitFor(()-> isFileUploadVisible(),
                    "the file upload panel did bot become visible", 2000);
        }
```

Reply:
good catch, thanks